### PR TITLE
Split NlSolver in multiple subsolvers

### PR DESCRIPTION
**Pull request by nafur**
The NlSolver started as one place for nonlinear reasoning, but has grown significantly since. This PR splits the NlSolver class into multiple smaller classes.

### CODE REVIEW

**Comment by nafur** [comment]
Benchmarks are currently running to check that this PR does not change solver behavior.

**Review by ajreynol — CHANGES_REQUESTED** [review]
I think we should use `nl/ext/` not `nl/nl/`.

File: src/theory/arith/nl/ext/shared_check_data.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+  /** Context-independent database of monomial information */
+  MonomialDb d_mdb;
+
+  // ( x*y, x*z, y ) for each pair of monomials ( x*y, x*z ) with common factors
+  std::map<Node, std::map<Node, Node> > d_mono_diff;
+  /** the set of monomials we should apply tangent planes to */
+  std::unordered_set<Node, NodeHashFunction> d_tplane_refine;
+
+  SharedCheckData(InferenceManager& im, NlModel& model)
```
Comment (ajreynol) [inline src/theory/arith/nl/ext/shared_check_data.h:56]: This should go to the top of the struct.

Comment (ajreynol) [reply src/theory/arith/nl/ext/shared_check_data.h:56]: Also move the constructor implementation to cpp.

File: src/theory/arith/nl/ext/shared_check_data.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#include "theory/arith/inference_manager.h"
+#include "theory/arith/nl/ext/monomial.h"
+#include "theory/arith/nl/nl_model.h"
+
+namespace CVC4 {
+namespace theory {
+namespace arith {
+namespace nl {
+
+struct SharedCheckData
```
Comment (ajreynol) [inline src/theory/arith/nl/ext/shared_check_data.h:30]: Can we call this class "NlExtState" or "ExtState"?

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/arith/nl/ext/ext_state.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+  /** Context-independent database of monomial information */
+  MonomialDb d_mdb;
+
+  // ( x*y, x*z, y ) for each pair of monomials ( x*y, x*z ) with common factors
+  std::map<Node, std::map<Node, Node> > d_mono_diff;
+  /** the set of monomials we should apply tangent planes to */
+  std::unordered_set<Node, NodeHashFunction> d_tplane_refine;
+
+  std::unique_ptr<CDProof> d_proof;
```
Comment (ajreynol) [inline src/theory/arith/nl/ext/ext_state.h:71]: I'm not sure a CDProof should live here.  What is the purpose of this?

File: src/theory/arith/nl/ext/ext_state.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  bool proofsEnabled() const;
+  bool addProof(Node expected,
+                PfRule id,
+                const std::vector<Node>& children,
+                const std::vector<Node>& args);
+  std::shared_ptr<ProofNode> getProof(Node expected);
+
+  void init(const std::vector<Node>& xts);
+
+  Node d_false = NodeManager::currentNM()->mkConst(false);
```
Comment (ajreynol) [inline src/theory/arith/nl/ext/ext_state.h:47]: I prefer these to be initialized in the constructor for consistency.

**Review by ajreynol — CHANGES_REQUESTED** [review]
A few more minor requests.

File: src/theory/arith/nl/ext/ext_state.h (hunk 3 above)
Comment (nafur) [reply src/theory/arith/nl/ext/ext_state.h]: Removed.

File: src/theory/arith/nl/ext/factoring_check.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+ private:
+  /** The inference manager that we push conflicts and lemmas to. */
+  InferenceManager& d_im;
+  /** Reference to the non-linear model object */
+  NlModel& d_model;
+  /** maps nodes to their factor skolems */
+  std::map<Node, Node> d_factor_skolem;
+
+  Node d_zero = NodeManager::currentNM()->mkConst(Rational(0));
```
Comment (ajreynol) [inline src/theory/arith/nl/ext/factoring_check.h:57]: Move to constructor.

File: src/theory/arith/nl/nonlinear_extension.h [hunk 6]
```diff
@@ -256,12 +260,17 @@ class NonlinearExtension
    * transcendental functions.
    */
   TranscendentalSolver d_trSlv;
+  ExtState d_extState;
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.h:263]: Document this.

File: src/theory/arith/nl/nonlinear_extension.h [hunk 7]
```diff
@@ -256,12 +260,17 @@ class NonlinearExtension
...
   /** The nonlinear extension object
    *
    * This is the subsolver responsible for running the procedure for
    * constraints involving nonlinear mulitplication, Cimatti et al., TACAS 2017.
    */
-  NlSolver d_nlSlv;
+  FactoringCheck d_factoringSlv;
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.h:269]: Update the documentation above these.

File: src/theory/arith/nl/ext/ext_state.h [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+#include "theory/arith/inference_manager.h"
+#include "theory/arith/nl/ext/monomial.h"
+#include "theory/arith/nl/nl_model.h"
+
+namespace CVC4 {
+namespace theory {
+namespace arith {
+namespace nl {
+
+struct ExtState
```
Comment (ajreynol) [inline src/theory/arith/nl/ext/ext_state.h:31]: Should maybe just make this a class.

**Review by ajreynol — CHANGES_REQUESTED** [review]
A few more requests.

**Review by ajreynol — APPROVED** [review]
LGTM, thanks a lot for doing this.

**Pull request merged**